### PR TITLE
Mock up of new protos for versioning proposal in #119.

### PR DESCRIPTION
**Pull request by ezyang**
This mocks up the new protos (without any supporting code) for the
key data structures that we need in the new versioning proposal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: onnx/namespace.proto [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  // These removals are backwards compatibility BREAKING changes; so
+  // whenever you remove a name, you must issue a new version of the
+  // operator namespace.
+  repeated string removed = 2; // namespace Operator
+
+  // The operator name bindings which were added in this version.
+  // Extra name bindings can be added to this field in a backwards
+  // compatible way without bumping the operator namespace version.
+  repeated OperatorBindingProto added = 3;
+}
```
Comment (linkerzhang) [inline onnx/namespace.proto]: question please. For semantic changes of an existing operator (say, Slice you mentioned :) ), will we only have it in "added" (overwrite way) or we also want to have it firstly in "removed" and then "added"?

Comment (ezyang) [reply onnx/namespace.proto]: Yep. So let's say you have `_OldSlice` and `_NewSlice`. Going in, you'll have a version of ONNX with `Slice` mapped to `_OldSlice`. Then, to move to NewSlice, you write this changelog entry:

```
ChangelogEntry {
  removed = "Slice"
  added = OperatorBinding {
    op_type = "Slice",
    op_def = "_NewSlice",
  }
}
```

We could introduce a modify field if this is a little oblique.

Comment (linkerzhang) [reply onnx/namespace.proto]: We may just have a "new" to contain both "added" and "changed". Say
ChangelogEntry {
  new = OperatorBinding {
    op_type = "Slice",
    op_def = "_NewSlice",
  }
}

In the "new", the key is always over-written, if the key exists, then it's removing the old one and change to new one, if the key does not exist, it's purely adding a new one, and "removed" contains keys totally removed.

Comment (linkerzhang) [reply onnx/namespace.proto]: the name "new" may be not good although :)

Comment (ezyang) [reply onnx/namespace.proto]: That's OK by me! My only wondering is whether or not we should be more explicit if an override occurs. "Adding" is a BC-compatible change, but "overriding" is not.

**Comment by gramalingam** [comment]
It seems like "op_def" is meant more to support an implementation, but doesn't have any formal or semantic content. (I don't have an objection to having it, but just checking if this name has any other meaning.)

Adding a descriptive "documentation" string would likely be more useful.

Now that we have this: this seems like a convenient place to start adding other meta-data about the operator (like its type-signature, etc.). Of course, we need to agree on a type-system first (about which there is a discussion elsewhere, I guess), but once we do so, would that be reasonable?

I guess this all boils to one question (discussed earlier): on the role of operator-registry in onnx/defs vs. explicitly capturing as much meta-data as possible in protobuf format.

**Comment by gramalingam** [comment]
Ok, please ignore part of my question, I read this before reading the new versioning proposal! At least I understand that op_def serves as a unique id. It's meaning is currently captured in the operator.md file, I guess. I still think it will be useful to expand the role of this definition to capture more meta-data about the operator right in the proto file, over time.

File: onnx/onnx.proto [hunk 2]
```diff
@@ -111,6 +111,11 @@ message ModelProto {
   // This field MUST be present.
   optional int64 ir_version = 1;
 
+  // The version of the operator namespace this model targets.  An operator
```
Comment (gramalingam) [inline onnx/onnx.proto]: So, I guess this is a MUST-be-present field?

Comment (ezyang) [reply onnx/onnx.proto]: Yep, definitely mandatory. But we could grandfather models missing this parameter by setting it to version 1.

**Comment by ezyang** [comment]
> I still think it will be useful to expand the role of this definition to capture more meta-data about the operator right in the proto file, over time.

Yes, I definitely think we're headed this direction. Actually, I would propose we put meta-data about operator definitions in a different proto file; one could definitely make use of the operator namespace proto without using any of the other type information, etc.

**Comment by donbox** [comment]
Sorry I'm late to the party.  I've now swapped in a bunch of context.

Here are my basic quesoins:

1. We need to anticipate organizations defining their own operator libraries for internal use that may or may not ever be published externally.  We will always have ONNX as the repository for operators that frameworks can claim support for, but at least in our world, we also know that there will be private operator repos that will have the same version requirements as we do.

2. I worry that the name mangling will have a lot of cognitive overhead AND requires a chain of change logs in order to find the trail of versions. If we just used SemVer per operator, we'd have a much simpler conceptual model AND the ability to have compatibility rules defined once for both graphs and ops.

3. It's not at all obvious wher ethe ChangeLogProto goes. Is it a separate top-level file format? If so how do we version that format? How does one acquire them and do I need all of them to complete the chain?

4. I saw you added op_version to model proto.  What's the format? SemVer? That assumes we are sticking with SemVer for ONNX for the duration, which we should talk about.  Our versioning discussions have been around the components, not the whole.  If we are IN ANY WAY refering to the release version, then we need to define the rules for versioning releases as a whole. Yikes!!

**Comment by ezyang** [comment]
Things to do:

1. Introduce vendor namespaces, so that vendors can also make use of the versioning mechanism for their private operators
2. Lay out the intent of defining machine readable format for the operator library

**Comment by ezyang** [comment]
Obsoleted.

**Pull request closed without merging**